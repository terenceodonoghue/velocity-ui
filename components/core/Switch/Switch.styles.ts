import { css, FunctionInterpolation, Theme } from '@emotion/react';
import { rgba, transitions as transition } from 'polished';

export const switchLabel = css`
  display: flex;
  align-items: center;
  justify-content: space-between;
  min-width: 36px;
  user-select: none;

  input {
    height: 0;
    margin: 0;
    width: 0;
    opacity: 0;
  }
`;

export const switchSlider: FunctionInterpolation<Theme> = ({
  palette,
  transitions,
}) => css`
  position: relative;
  border-radius: 10px;
  height: 20px;
  width: 36px;
  background-color: ${rgba(palette.neutral[600], 0.4)};
  cursor: pointer;
  ${transition(
    ['background-color', 'opacity'],
    `${transitions.duration.standard}ms ${transitions.easing.sharp}`,
  )};

  input:disabled + & {
    opacity: 0.4;
  }

  input:checked:enabled + & {
    background-color: ${palette.ui.green};
  }

  &::before {
    position: absolute;
    content: '';
    left: 2px;
    top: 2px;
    border-radius: 50%;
    height: 16px;
    width: 16px;
    transition: transform ${transitions.duration.standard}ms
      ${transitions.easing.sharp};

    input:checked + & {
      transform: translateX(16px);
    }
  }
`;

export const switchText: FunctionInterpolation<Theme> = ({
  palette,
  typography,
}) => css`
  display: block;
  margin: 0 58px 0 0;
  font-weight: ${typography.fontWeight.medium};

  &:nth-of-type(2) {
    color: ${palette.neutral[600]};
    font-weight: ${typography.fontWeight.regular};
  }
`;
