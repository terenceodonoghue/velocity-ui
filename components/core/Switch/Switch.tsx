import React, { forwardRef } from 'react';
import * as css from './Switch.styles';
import type { SwitchProps } from './Switch.types';

const Switch = forwardRef<HTMLInputElement, SwitchProps>(
  ({ primary, secondary, ...props }, ref) => (
    <label css={css.switchLabel} data-testid="switch">
      {(primary || secondary) && (
        <>
          {primary && (
            <span css={css.switchText} data-testid="switch-primary">
              {primary}
            </span>
          )}
          {secondary && (
            <span css={css.switchText} data-testid="switch-secondary">
              {secondary}
            </span>
          )}
        </>
      )}
      <input data-testid="switch-input" ref={ref} type="checkbox" {...props} />
      <span css={css.switchSlider} />
    </label>
  ),
);

export default Switch;
