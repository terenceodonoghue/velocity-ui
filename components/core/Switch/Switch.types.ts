import { InputHTMLAttributes } from 'react';

export interface SwitchProps extends InputHTMLAttributes<HTMLInputElement> {
  primary?: string;
  secondary?: string;
}
