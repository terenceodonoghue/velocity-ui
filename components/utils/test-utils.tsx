import React, { FunctionComponent, ReactElement } from 'react';
import { render, RenderOptions, RenderResult } from '@testing-library/react';
import ThemeProvider from '~/components/core/ThemeProvider';

const Providers: FunctionComponent = ({ children }) => {
  return <ThemeProvider>{children}</ThemeProvider>;
};

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
): RenderResult => render(ui, { wrapper: Providers, ...options });

export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';

export { customRender as render };
