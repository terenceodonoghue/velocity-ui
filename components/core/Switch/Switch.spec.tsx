import React from 'react';
import faker from 'faker';
import { render, userEvent } from 'test-utils';
import Switch from './Switch';

describe('Switch', () => {
  it('renders primary label', () => {
    const content = faker.hacker.phrase();
    const { container, getByTestId, queryByTestId } = render(
      <Switch primary={content} />,
    );

    expect(getByTestId('switch-primary')).toHaveTextContent(content);
    expect(queryByTestId('switch-secondary')).toBeNull();
    expect(container).toMatchSnapshot();
  });

  it('renders secondary label', () => {
    const content = faker.hacker.phrase();
    const { container, getByTestId, queryByTestId } = render(
      <Switch secondary={content} />,
    );

    expect(queryByTestId('switch-primary')).toBeNull();
    expect(getByTestId('switch-secondary')).toHaveTextContent(content);
    expect(container).toMatchSnapshot();
  });

  it('renders primary and secondary labels', () => {
    const content = faker.hacker.phrase();
    const { container, getByTestId } = render(
      <Switch primary={content} secondary={content} />,
    );

    expect(getByTestId('switch-primary')).toHaveTextContent(content);
    expect(getByTestId('switch-secondary')).toHaveTextContent(content);
    expect(container).toMatchSnapshot();
  });

  it('is unchecked by default', () => {
    const { getByTestId } = render(<Switch />);

    expect(getByTestId('switch-input')).not.toBeChecked();
  });

  it('is checked when clicked', () => {
    const { getByTestId } = render(<Switch />);

    userEvent.click(getByTestId('switch'));

    expect(getByTestId('switch-input')).toBeChecked();
  });
});
