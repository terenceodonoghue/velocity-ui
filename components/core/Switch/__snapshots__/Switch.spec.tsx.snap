// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch renders primary and secondary labels 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  min-width: 36px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-0 input {
  height: 0;
  margin: 0;
  width: 0;
  opacity: 0;
}

.emotion-1 {
  display: block;
  margin: 0 58px 0 0;
  font-weight: 500;
}

.emotion-1:nth-of-type(2) {
  color: #8798ad;
  font-weight: 400;
}

.emotion-3 {
  position: relative;
  border-radius: 10px;
  height: 20px;
  width: 36px;
  background-color: rgba(135,152,173,0.4);
  cursor: pointer;
  -webkit-transition: background-color 300ms cubic-bezier(0.4, 0, 0.6, 1),opacity 300ms cubic-bezier(0.4, 0, 0.6, 1);
  transition: background-color 300ms cubic-bezier(0.4, 0, 0.6, 1),opacity 300ms cubic-bezier(0.4, 0, 0.6, 1);
}

input:disabled+.emotion-3 {
  opacity: 0.4;
}

input:checked:enabled+.emotion-3 {
  background-color: #33ac2e;
}

.emotion-3::before {
  position: absolute;
  content: '';
  left: 2px;
  top: 2px;
  border-radius: 50%;
  height: 16px;
  width: 16px;
  -webkit-transition: -webkit-transform 300ms cubic-bezier(0.4, 0, 0.6, 1);
  transition: transform 300ms cubic-bezier(0.4, 0, 0.6, 1);
}

input:checked+.emotion-3::before {
  -webkit-transform: translateX(16px);
  -moz-transform: translateX(16px);
  -ms-transform: translateX(16px);
  transform: translateX(16px);
}

<div>
  <label
    class="emotion-0"
    data-testid="switch"
  >
    <span
      class="emotion-1"
      data-testid="switch-primary"
    >
      You can't input the array without navigating the redundant GB alarm!
    </span>
    <span
      class="emotion-1"
      data-testid="switch-secondary"
    >
      You can't input the array without navigating the redundant GB alarm!
    </span>
    <input
      data-testid="switch-input"
      type="checkbox"
    />
    <span
      class="emotion-3"
    />
  </label>
</div>
`;

exports[`Switch renders primary label 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  min-width: 36px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-0 input {
  height: 0;
  margin: 0;
  width: 0;
  opacity: 0;
}

.emotion-1 {
  display: block;
  margin: 0 58px 0 0;
  font-weight: 500;
}

.emotion-1:nth-of-type(2) {
  color: #8798ad;
  font-weight: 400;
}

.emotion-2 {
  position: relative;
  border-radius: 10px;
  height: 20px;
  width: 36px;
  background-color: rgba(135,152,173,0.4);
  cursor: pointer;
  -webkit-transition: background-color 300ms cubic-bezier(0.4, 0, 0.6, 1),opacity 300ms cubic-bezier(0.4, 0, 0.6, 1);
  transition: background-color 300ms cubic-bezier(0.4, 0, 0.6, 1),opacity 300ms cubic-bezier(0.4, 0, 0.6, 1);
}

input:disabled+.emotion-2 {
  opacity: 0.4;
}

input:checked:enabled+.emotion-2 {
  background-color: #33ac2e;
}

.emotion-2::before {
  position: absolute;
  content: '';
  left: 2px;
  top: 2px;
  border-radius: 50%;
  height: 16px;
  width: 16px;
  -webkit-transition: -webkit-transform 300ms cubic-bezier(0.4, 0, 0.6, 1);
  transition: transform 300ms cubic-bezier(0.4, 0, 0.6, 1);
}

input:checked+.emotion-2::before {
  -webkit-transform: translateX(16px);
  -moz-transform: translateX(16px);
  -ms-transform: translateX(16px);
  transform: translateX(16px);
}

<div>
  <label
    class="emotion-0"
    data-testid="switch"
  >
    <span
      class="emotion-1"
      data-testid="switch-primary"
    >
      The SMS program is down, synthesize the redundant feed so we can synthesize the SQL hard drive!
    </span>
    <input
      data-testid="switch-input"
      type="checkbox"
    />
    <span
      class="emotion-2"
    />
  </label>
</div>
`;

exports[`Switch renders secondary label 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  min-width: 36px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.emotion-0 input {
  height: 0;
  margin: 0;
  width: 0;
  opacity: 0;
}

.emotion-1 {
  display: block;
  margin: 0 58px 0 0;
  font-weight: 500;
}

.emotion-1:nth-of-type(2) {
  color: #8798ad;
  font-weight: 400;
}

.emotion-2 {
  position: relative;
  border-radius: 10px;
  height: 20px;
  width: 36px;
  background-color: rgba(135,152,173,0.4);
  cursor: pointer;
  -webkit-transition: background-color 300ms cubic-bezier(0.4, 0, 0.6, 1),opacity 300ms cubic-bezier(0.4, 0, 0.6, 1);
  transition: background-color 300ms cubic-bezier(0.4, 0, 0.6, 1),opacity 300ms cubic-bezier(0.4, 0, 0.6, 1);
}

input:disabled+.emotion-2 {
  opacity: 0.4;
}

input:checked:enabled+.emotion-2 {
  background-color: #33ac2e;
}

.emotion-2::before {
  position: absolute;
  content: '';
  left: 2px;
  top: 2px;
  border-radius: 50%;
  height: 16px;
  width: 16px;
  -webkit-transition: -webkit-transform 300ms cubic-bezier(0.4, 0, 0.6, 1);
  transition: transform 300ms cubic-bezier(0.4, 0, 0.6, 1);
}

input:checked+.emotion-2::before {
  -webkit-transform: translateX(16px);
  -moz-transform: translateX(16px);
  -ms-transform: translateX(16px);
  transform: translateX(16px);
}

<div>
  <label
    class="emotion-0"
    data-testid="switch"
  >
    <span
      class="emotion-1"
      data-testid="switch-secondary"
    >
      You can't quantify the interface without parsing the neural ADP feed!
    </span>
    <input
      data-testid="switch-input"
      type="checkbox"
    />
    <span
      class="emotion-2"
    />
  </label>
</div>
`;
